### PR TITLE
Smoothnesss and more safe obstacle avoidance

### PR DESCRIPTION
**Pull request by artofnothingness**
#41

### CODE REVIEW

File: src/controller.cpp [hunk 1]
```diff
@@ -57,6 +57,8 @@ geometry_msgs::msg::TwistStamped Controller::computeVelocityCommands(
   nav2_core::GoalChecker * goal_checker)
 {
   std::lock_guard<std::mutex> lock(*parameters_handler_->getLock());
+
+  auto start = std::chrono::high_resolution_clock::now();
```
Comment (artofnothingness) [inline src/controller.cpp:61]: let leave it as is till release

Comment (SteveMacenski) [reply src/controller.cpp:61]: Sure thing

File: src/critics/smoothness_critic.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  using namespace xt::placeholders;  // NOLINT
+  
+  if (!enabled_) {
+    return;
+  }
+
+  auto vx = data.state.getControlVelocitiesVX();
+  auto wz = data.state.getControlVelocitiesWZ();
+
+  auto vx_diff = xt::abs(xt::view(vx, xt::all(), xt::range(1, _)) - xt::view(vx, xt::all(), xt::range(0, -1)));
```
Comment (SteveMacenski) [inline src/critics/smoothness_critic.cpp]: Forgive my ignorance, but on L33 you use `wz` and then on L35/36/42/43 I see a ix of `vx` and `wz`'s when I think they should be `vy` and such. It seems like maybe a copy paste error?

**Comment by SteveMacenski** [comment]
Did you notice this to help?

File: src/critics/smoothness_critic.cpp (hunk 2 above)
Comment (artofnothingness) [reply src/critics/smoothness_critic.cpp]: fixed

**Comment by artofnothingness** [comment]
i didn't notice much wobblings, especially with point cost. Check it plz too

**Comment by SteveMacenski** [comment]
I'll be honest, I have a hard time seeing the wobble / jerking in Rviz even before this. How are you visualizing this to see that behavior?

**Comment by artofnothingness** [comment]
I mostly look at trajectory visualization and it's stability

**Comment by SteveMacenski** [comment]
Got it- Lets see what Pradheep says. I also went through to scavenge for parts of old robot projects to put together a platform. It won't be overnight, but I should have something to test with in coming weeks physically.

**Review by SteveMacenski — APPROVED** [review]
Approve the code, but I think we need testing on physical robot to tell the degree to which this solves or partially solves the problem

**Comment by SteveMacenski** [comment]
@artofnothingness I would really like this tested on hardware before we add. I really don't actually see this in my testing as helping much if any. I think what you're seeing is that since the time_steps were reduced low enough it looks more stable, but if you increase back to `30` timesteps for 3 seconds in the future, it seems just as wobbly as before, to my eye at least.  

I see alot more stability offered by changing `temperature` from 0.25 -> 0.5. But it could be my eye. I think 1.2s is too short though, I think 2.5s would be a dead minimum for practical use when a robot is moving at a reasonable speed like `0.7m/s` which is approximately human walking speed (and still pretty slow for a robot)

**Comment by artofnothingness** [comment]
@SteveMacenski probably it requires weights tweaks. it basically  d | u | that you mentioned before it should help

**Comment by SteveMacenski** [comment]
I tried `0.2` (default from readme), `1.0`,  `5.0`, `50.0` and none helped. In fact, the higher it went, it got extremely unstable, please test yourself. I can't take a screen video for some reason on 22.04, it all shows up black and some documentation claims its for security purposes but I'm not sure the best way around it right now. 

I think this PR needs to be reverted. It makes the behavior much worse. I can show with it removed with any weight from `0.2` - `50.0` is better without it. This makes the robot paths very drunken and irregular.

**Comment by SteveMacenski** [comment]
And that kind of makes sense too - since we're penalizing the change in velocity for each timestep, so it makes it so a robot wants to continue moving in the direction it was previously moving. Thus, it will really resist straightening out if it was already turning. Since we randomly sample each iteration, there's going to always be random little perturbations creating these oscillations. The more we penalize changes, the more extreme the oscillations will become 
 
That is proved more if you also increase the cost power to `2`, it goes a bit nuts. For small values of `weight` it seems fine, but I'm not sure its helping. It seems to me why it seems 'fine' is just because its not big enough to overpower the other critics. The lower I make it, the better behavior I see, until `0.0`, so I think it isn't helping. I see little, if any, change in behavior from `0.0` - `0.4` and just not including the critic at all. From `0.6+`, it get notably worse. 

I think we're penalizing the slightly wrong thing here.

I tried modifying it to penalizing from the `data.state.speed` and just the 1st control against the speed instead which didn't help. Also tried error w.r.t. a zero deviation (e.g. `0` in angular and `max_speed` in Vx) and that did not help either.

**Comment by artofnothingness** [comment]
I did see the same with high weights. I just guess that low cost should help a little, even if we can't notice that, but now i have doubts about that

**Comment by SteveMacenski** [comment]
Also just tried `sum` the changes instead of averaging them - no dice. I thought that might be the solution

**Pull request merged**